sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -132,9 +134,12 @@ object NestedColumnAliasing {
 
         // If all nested fields of `attr` are used, we don't need to introduce new aliases.
         // By default, ColumnPruning rule uses `attr` already.
+        // Note that we need to remove cosmetic variations first, so we only count a
+        // nested field once.
         if (nestedFieldToAlias.nonEmpty &&
-            nestedFieldToAlias
-              .map { case (nestedField, _) => totalFieldNum(nestedField.dataType) }
+            nestedFields.map(_.canonicalized)
```

Comment:
Is this the main difference from 3.0, `dedupNestedFields` -> `nestedFields`?

Reply:
Yeah, for the change in `NestedColumnAliasing`. Another difference is test. One test in master branch cannot pass in branch-3.0.

Reply:
Yes. Test part looked correct because it's a subset. For this part, it looks a little different and needs more validation.

Reply:
Thanks for check. Yes, there is a bit difference between master and branch-3.0 here. So no `dedupNestedFields` in branch-3.0.

Reply:
Ah, I missed this part. I think the added test still fails if we don't have this change. Is this correct?

Reply:
Yes. It does. The new test case still validate this patch.

Reply:
Yes, this test fails in current branch-3.0.

Reply:
Nice.